research/job_referencing/index.md

```diff
@@ -0,0 +1,217 @@
+When running, as an example, the COPR build for the latest dependent package in the list (C in our example), at the end of the handler we can create the COPR build tasks for package A (if any).
+This can be done for all the kinds of job we directly start at [this lines](https://github.com/packit/packit-service/blob/16236087fb8071c83b3802464f3af504f4fb933e/packit_service/worker/jobs.py#L464-L497).
+
+We can change `bodhi` jobs in a way that they will do nothing, apart from starting new `bodhi` tasks for other packages, unless they belong to a package not required by another package. Ideally only package A, in this example, should do a `bodhi` update if I am not wrong. Or at least we should be able to skip doing something in this job during the chain of dependencies.
+
+I think we don't need changes in the database schema to be able to implement this solution.
+We need to add the `depends` or `depends_on` key to the schema. I slightly prefer the `depends` key, but not strong opinions on that.
+
+### Request new feature: support side tag for multi package update
+
+This is the [RFE](https://github.com/packit/packit/issues/1870).
+
+I am not really sure I am getting the point here. But they are talking about a multi-package feature and for this reason I assume they need to use the monorepo syntax (otherwise we have no way to reference the packages).
+If they can and want to use the monorepo syntax then I think that the package dependencies solution for the monorepos would solve their problems too. Or am I wrong?
+
+## Drawbacks
+
+I see just a drawback in the previous solutions, the code is getting more and more complex to read (I mean to grasp the execution order) and test.
+I think `Celery` was well suited for tasks which didn't depend on each other, now that dependencies are growing quickly I feel it is like a limit. I would like to be able to have communication between tasks and suspend the execution without exiting.
+I think that a code using `asyncio`, as an example, could be more readable and testable.
+
+Let's take as an example the simple dependency we have between COPR builds and SRPM builds
+
+[This code](https://github.com/packit/packit-service/blob/16236087fb8071c83b3802464f3af504f4fb933e/packit_service/worker/handlers/copr.py#L136-L203) is performing both builds. But it is not straightforward to know that this code is called twice, because there are two `CoprBuildStartEvent`(s) sent by the COPR server. The execution flow is not obvious and not easily testable, even though this is a really simple example.
+
+I would like to be able to write something similar using `asyncio`.
+As far as I can understand it [temporalio](https://docs.temporal.io) would give us this flexibility. For this reason I wrote down some _python pseudo code_ using `temporalio`. I don't really know it so I can be missing something.
```

Comment:
(Just asking.) Have you considered any other solution or is this the only one the might be suitable for us?

Reply:
I went through all the workflow engines listed in the research, from my point of view temporal.io is the only one that will allow us to write *linear code* and, if I get it right, it will take the burden to manage the sleeping tasks in a way completely transparent to us. But I may be wrong, I don't know any of them.

Reply:
Thanks!